### PR TITLE
Submit EIP-3525 for review

### PR DESCRIPTION
**Pull request by will-edge**
All modifications according to last PR, files for reference implementations.

### CODE REVIEW

File: assets/eip-3525/contracts/extensions/IERC3525Metadata.sol [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: CC0-1.0
+pragma solidity ^0.8.0;
+
+import "../IERC3525.sol";
+import "../openzeppelin/IERC721Metadata.sol";
```
Comment (MicahZoltu) [inline assets/eip-3525/contracts/extensions/IERC3525Metadata.sol:5]:
```suggestion
import "../IERC3525.sol";
import "IERC721Metadata.sol";
```
Rather than have an external dependency on a specific implementation, it should be left as an exercise for the reader to find an implementation.

File: assets/eip-3525/contracts/IERC3525.sol [hunk 2]
```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: CC0-1.0
+pragma solidity ^0.8.0;
+
+import "./openzeppelin/IERC165.sol";
+import "./openzeppelin/IERC721.sol";
```
Comment (MicahZoltu) [inline assets/eip-3525/contracts/IERC3525.sol:5]:
```suggestion
import "IERC165.sol";
import "IERC721.sol";
```
Finding an implementation for dependencies is left as an exercise for the reader.

File: assets/eip-3525/LICENSE [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+CC0 1.0 Universal
+
+    CREATIVE COMMONS CORPORATION IS NOT A LAW FIRM AND DOES NOT PROVIDE
+    LEGAL SERVICES. DISTRIBUTION OF THIS DOCUMENT DOES NOT CREATE AN
+    ATTORNEY-CLIENT RELATIONSHIP. CREATIVE COMMONS PROVIDES THIS
+    INFORMATION ON AN "AS-IS" BASIS. CREATIVE COMMONS MAKES NO WARRANTIES
+    REGARDING THE USE OF THIS DOCUMENT OR THE INFORMATION OR WORKS
+    PROVIDED HEREUNDER, AND DISCLAIMS LIABILITY FOR DAMAGES RESULTING FROM
+    THE USE OF THIS DOCUMENT OR THE INFORMATION OR WORKS PROVIDED
+    HEREUNDER.
```
Comment (MicahZoltu) [inline assets/eip-3525/LICENSE:12]: Everything in this repository is CC0, so this file can be left out.  This is especially true with the comment at the top of each of the .sol files.

File: assets/eip-3525/README.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+- Solidity 0.8
+
+### Setup
+
+Run `npm isntall` in the root directory
+
+## Unit Tests
+
+Run `npm test` to run the unit tests
```
Comment (MicahZoltu) [inline assets/eip-3525/README.md:17]: This is redundant with the EIP and should be deleted.
```suggestion
```

File: assets/eip-3525/contracts/ERC3525Upgradeable.sol [hunk 5]
```diff
@@ -0,0 +1,718 @@
...
+import "./IERC3525Receiver.sol";
+import "./extensions/IERC3525Metadata.sol";
+import "./openzeppelin/ContextUpgradeable.sol";
+import "./openzeppelin/ERC165Upgradeable.sol";
+import "./openzeppelin/IERC721Enumerable.sol";
+import "./openzeppelin/IERC721Metadata.sol";
+import "./openzeppelin/IERC721Receiver.sol";
+import "./utils/base64.sol";
+import "./utils/StringConvertor.sol";
+import "hardhat/console.sol";
```
Comment (MicahZoltu) [inline assets/eip-3525/contracts/ERC3525Upgradeable.sol:14]: This feels like it is out of scope for a reference implementation.  The reference implementation should do the bare minimum required to show how one might create a very simple implementation of this standard, and since this isn't an upgradability standard this whole file is out of scope and adds unnecessary complexity to the reference implementation.

File: assets/eip-3525/test/test.ts [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+      expect(await t.erc3525.isApprovedForAll(t.owner, approval.address)).to.eq(
+        true
+      );
+      expect(
+        await t.erc3525.isApprovedForAll(
+          t.owner,
+          "0x000000000000000000000000000000000000dEaD"
+        )
+      ).to.eq(false);
+    });
```
Comment (MicahZoltu) [inline assets/eip-3525/test/test.ts:131]: Ideally, the tests should be converted to a more portable format.  A good example of this is an array of test cases with expected outputs, exceptions, logs, and post-conditions for each.  At the moment, this test file comes with a lot of JS specific stuff and includes deployment code which is out of scope for a good set of test cases.  Note, the section is called "Test Cases" not "Test Infrastructure".  😀

File: assets/eip-3525/LICENSE (hunk 3 above)
Comment (Pandapip1) [reply assets/eip-3525/LICENSE:12]: > This is especially true with the comment at the top of each of the .sol files.

Actually, it's required by the Solidity style guide! Please keep them in.

File: assets/eip-3525/LICENSE [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+the greatest extent permissible under applicable law.
+c. Affirmer disclaims responsibility for clearing rights of other persons
+that may apply to the Work or any use thereof, including without
+limitation any person's Copyright and Related Rights in the Work.
+Further, Affirmer disclaims responsibility for obtaining any necessary
+consents, permissions or other rights required for any use of the
+Work.
+d. Affirmer understands and acknowledges that Creative Commons is not a
+party to this document and has no duty or obligation with respect to
+this CC0 or use of the Work.
```
Comment (Pandapip1) [inline assets/eip-3525/LICENSE:121]:
```suggestion
```

File: assets/eip-3525/contracts/ERC3525SlotApprovableUpgradeable.sol [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        uint256 slot = ERC3525Upgradeable.slotOf(tokenId_);
+        return (
+            operator_ == owner ||
+            getApproved(tokenId_) == operator_ ||
+            ERC3525Upgradeable.isApprovedForAll(owner, operator_) ||
+            ERC3525SlotApprovableUpgradeable.isApprovedForSlot(owner, slot, operator_)
+        );
+    }
+
+}
```
Comment (Pandapip1) [inline assets/eip-3525/contracts/ERC3525SlotApprovableUpgradeable.sol:84]:
```suggestion
}

```

File: assets/eip-3525/contracts/ERC3525SlotEnumerableUpgradeable.sol [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        uint256 tokenIndex = slotData.slotTokens[tokenId_];
+
+        slotData.slotTokens[tokenIndex] = lastTokenId;
+        _slotTokensIndex[slot_][lastTokenId] = tokenIndex;
+
+        delete _slotTokensIndex[slot_][tokenId_];
+        slotData.slotTokens.pop();
+    }
+
+}
```
Comment (Pandapip1) [inline assets/eip-3525/contracts/ERC3525SlotEnumerableUpgradeable.sol:118]:
```suggestion
}

```

File: assets/eip-3525/contracts/IERC3525Receiver.sol [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+     * @param _operator The address which triggered the transfer
+     * @param _fromTokenId The token id to transfer value from
+     * @param _toTokenId The token id to transfer value to
+     * @param _value The transferred value
+     * @param _data Additional data with no specified format
+     * @return `bytes4(keccak256('onERC3525Received(address,uint256,uint256,uint256,bytes)'))` 
+     *  unless the transfer is rejected.
+     */
+    function onERC3525Received(address _operator, uint256 _fromTokenId, uint256 _toTokenId, uint256 _value, bytes calldata _data) external returns (bytes4);
+}
```
Comment (Pandapip1) [inline assets/eip-3525/contracts/IERC3525Receiver.sol:26]:
```suggestion
}

```

File: EIPS/eip-3525.md [hunk 11]
```diff
@@ -546,44 +570,53 @@ This is the "ERC-3525 Metadata JSON Schema for tokenURI(uint)" referenced above.
...
 
 ## Backwards Compatibility
 
-As mentioned in the beginning, ERC-3525 is an extension interface of ERC-721, and therefore 100% compatible with the latter.
+As mentioned in the beginning, this EIP is an compatible standard of EIP-721.
+
+
+## Reference Implementation
+
+- [EIP-3525 implementation by SOLV Protocol](../assets/eip-3525)
```
Comment (Pandapip1) [inline EIPS/eip-3525.md:609]: This needs to link to the individual files. There are no directory listings.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

File: assets/eip-3525/LICENSE (hunk 3 above)
Comment (MicahZoltu) [reply assets/eip-3525/LICENSE:12]: This repository is under no compulsion to follow the Solidity style guide, especially when it is being dumb and requiring redundant files.  😀 

I would want to hear a far more compelling reason to keep this than, "because someone on the internet thinks it should be included in complete projects".

**Comment by Pandapip1** [comment]
I, for some reason, thought Git wouldn't recognize empty files. My bad! Please delete them.

**Review by Pandapip1 — APPROVED** [review]
LGTM

**Comment by Pandapip1** [comment]
Please close and re-open this PR in order for this to merge.
